### PR TITLE
User/ghaack/cartridge

### PR DESCRIPTION
**Pull request by grantbreaksthings**
Hey Blazer,

I was inspired by this project and wanted to contribute to it. I can see that you're starting to add compatibility for games beyond Tetris and I think some of my code will be useful for that. Here are some major changes I've made

- Added a Cartridge class. This contains virtual functions for reading and writing to the cartridge, as well as other metadata that all cartridges share.
- Added an emulated MBC1 based cartridge. This has support for multiple ROM and RAM banks. It should work for getting the teensy.gb working with Super Mario Land.
- Added an emulated No MBC based cartridge. This has support for unbanked ROM and RAM and should work with Tetris.
- Added numerous helper functions and lookup tables for parsing the type of cartridge that a ROM uses. For example, lookupRomSize, lookupCartType, lookupRamSize, lookupMbcType, etc.
- Added SD card support. ROMs can be loaded from an SD card instead of a .gitignored cpp file. (Tested on Teensy 4.1)
- Memory is no longer a static object. I did this because the Memory object contains a Cartridge object, and the Cartridges cannot be static. Cartridges are virtual and the type of cartridge that gets created is decided at runtime based on the ROM file that is loaded. The CPU and PPU are initialized with a handle to the Memory object.
- Simulated GameBoy memory is no longer a contiguous block but is instead broken up into different regions (vram, wram, oam, etc). This was done to prevent declaring memory regions twice since external RAM exists in the Cartridge class instead of the Memory class. It's also a bit easier to understand.

While this does build, I must admit that many of these features are untested as I'm still waiting on a ER-TFT040-2 from BuyDisplay. Even though some features are untested, I think the code might still be useful.

### CODE REVIEW

**Comment by blazer82** [comment]
Hey @granthaack,

Thank you very much for your contribution. I've seen you also forked and adapted my display driver 😊

I will have an in-depth look at your code as soon as I get to it. I'm pretty sure I can use a lot of it for cartridge handling as I pretty much neglected this part so far. And if this is going to help me getting Super Mario Land running it's even better because I've been running into a lot of issues with this one.

As you've probably seen, I've been busy, too. That's why your master is out of sync with mine. I'll try to resolve this, though.

**Comment by blazer82** [comment]
@granthaack thanks for keeping this PR updated. Have you been able to confirm that it works on the hardware yet?

I'll be focussing on proper sound implementation next, so I probably won't be merging this right now. Also, I wanna have a good look at the dynamic memory allocation you introduced as well as the new dependency on the Teensy 4.1.

**Comment by grantbreaksthings** [comment]
I have not tested it on actual hardware yet. My first screen from BuyDisplay got lost in the mail, I rush ordered a second one and it should be coming today. I'll assemble the display driver and test everything ASAP.

I'll keep this branch up to date while you work on sound.

**Comment by grantbreaksthings** [comment]
IMG_20200901_190357
I have confirmed that these changes work on actual hardware.

**Comment by blazer82** [comment]
@granthaack very nice! Also my first time seeing the board in red 😃

I'm making good progress on the APU with an already very recognizable Tetris melody. Once this is mostly done, I'll try to integrate your code. I also bought myself a Teensy 4.1 so I can test your code, too 😊

File: lib/CPU/CPU.cpp [hunk 1]
```diff
@@ -36,12 +36,35 @@
...
+// HL Register
+// Bit 0 - 7:   Gen Purpose L Register
+// Bit 8 - 15:  Gen Purpise H Register
+uint16_t CPU::HL = 0x014D;
+
+// Stack Pointer
+uint16_t CPU::SP = 0xFFFE;
+
+// Program Counter
+uint16_t CPU::PC = PC_START;
```
Comment (blazer82) [inline lib/CPU/CPU.cpp]: Thanks for improving the documentation, I appreciate it!

I'm thinking about changing the registers into union types as I did with the APU registers but that's definitely for another time! First I wanna get this one merged.

File: lib/Memory/Memory.cpp [hunk 2]
```diff
@@ -27,48 +27,48 @@
 
 #define MAX(a, b) (((a) > (b)) ? (a) : (b))
 
-bool Memory::mbc1_mode = 0;
-uint8_t Memory::romBank = 0;
-
-uint8_t Memory::memory[0x8000] = {0};
-
-void Memory::writeByteInternal(const uint16_t location, const uint8_t data, const bool internal) {
+void Memory::writeByteInternal(uint16_t location, uint8_t data, bool internal) {
```
Comment (blazer82) [inline lib/Memory/Memory.cpp]: Why are the first two params not `const` anymore? You're not going to reassign them anywhere, right?

File: lib/Memory/Memory.cpp [hunk 3]
```diff
@@ -113,56 +113,130 @@ void Memory::writeByteInternal(const uint16_t location, const uint8_t data, cons
...
+                Serial.println("Illegal write operation on memory!");
+                Serial.printf("\tAttempted write of 0x%x to 0x%x\n", data, location);
             }
-
             break;
     }
 }
 
-void Memory::writeByte(const uint16_t location, const uint8_t data) { writeByteInternal(location, data, false); }
+void Memory::writeByte(uint16_t location, uint8_t data) { writeByteInternal(location, data, false); }
```
Comment (blazer82) [inline lib/Memory/Memory.cpp]: Same here with the `const`

**Review by blazer82 — COMMENTED** [review]
You did some awesome work with this PR, thanks again! This will really help push the project forward. I will test and merge this within the next couple of days.

I also left some comments.

Also: I added some CI to the repo, mostly to enforce proper code formatting and make sure the code builds. Ironically, the current master won't build because of my incomplete cartridge implementation. Your PR may fix this 😊

File: lib/Memory/Memory.cpp (hunk 2 above)
Comment (grantbreaksthings) [reply lib/Memory/Memory.cpp]: My bad, this should still be const

File: lib/Memory/Memory.cpp (hunk 3 above)
Comment (grantbreaksthings) [reply lib/Memory/Memory.cpp]: This too

**Comment by grantbreaksthings** [comment]
Yeah moving away from static was a terrible idea, but a learning experience!
I have merged your upstream APU and joypad changes and confirmed that they worked.

**Comment by blazer82** [comment]
@granthaack ~so, I checked out your branch and tried running it on my Teensy 4.1 and I can't seem to get past the `Initializing SD card...` part.~

~I tried inserting an SD card formatted with FAT32, FAT16 and FAT12 and a `tetris.gb` file in its root as specified by `Cartridge::begin("tetris.gb")`. Any idea what I might be doing wrong?~

**Comment by blazer82** [comment]
@granthaack I got it! Used the wrong partition format 😅

**Comment by blazer82** [comment]
I confirmed that this code works - awesome! 😃

The APU suddenly seems to be missing in this PR, though. I suppose you accidentally removed it during a cleanup. If you fix this and also adapt the latest changes from the masters `main.cpp` I'll merge this and maybe clean it up somewhat afterwards.

If you want to you can also fix the linting errors but I'll happily do this after merging 😊

**Comment by grantbreaksthings** [comment]
The code works at this point and the APU and Joypad are included, I'm trying to learn how to use .clang-format to lint everything so it will pass the build checks.

**Comment by grantbreaksthings** [comment]
Well that was an adventure. I've gotten the build to pass and the code still works.

**Comment by blazer82** [comment]
OUTSTANDING WORK 👏👏👏

I will need to add some APU related logic to Memory.cpp and I'm also going to remove the clang-format python script from the repository again because if you configure your editor to automatically format your files on save this shouldn't be required.

Let me tell you, though: I'm so happy with what I'll be able to do with this PR merged as I wanted to integrate SD card support for a while but it wasn't at the top of my list - and now it's here! Not to mention the extended cartridge support 😃

**Comment by grantbreaksthings** [comment]
My bad, that python file shouldn't have gotten committed. Thanks for your help merging this. MBC1 should be fully functional, it will load (some) of the title screen for Super Mario Land and a couple other MBC1 games I've tried (baseball, tennis). The groundwork is there for MBC2 as well, but it's untested.

**Comment by blazer82** [comment]
Yes, you were right with the first version of you comment: the PPU definitely needs to be worked on as I merely implemented all the features required to run Tetris. Some important features like windows and scrolling are currently missing entirely.

What's somewhat more of a concern to me right now is that Super Mario Land still won't respond to button inputs. And also: Blargg's CPU tests, although passing when ran individually, derail during execution of test 3. That one has been a mystery to me for a long time now and I was kind of hoping the issue was with my improvised MBC implementation.

On a side note: have you been able to wire up your version of the hardware and at least have some fun playing Tetris with sound yet? Let me know if you need some help there! I do plan on adding a rudimentary wiring guide to the readme but it's not ready yet.

**Comment by grantbreaksthings** [comment]
I see that issue as well. I got the Super Mario Land title screen to render properly with some additions to the PPU but it won't respond to button inputs for me either. The same problem seems to happen with Baseball as well. I've also tried Tennis and Alleyway. Tennis gets passed the title screen but starts to fall apart during gameplay and Alleyway doesn't load at all. 

My PPU changes are in user/ghaack/tile-upgrades, feel free to pull that in to master if you like. I'm going to focus on getting Blargg's various tests to pass. The Teensy is a difficult platform to debug (no SWD/JTAG) so this is probably the quickest way to getting more games working.

I haven't gotten the sound working yet, I can see that you're outputting the audio channels on separate pins and I'm not sure how to mix them.

**Comment by blazer82** [comment]
I got the CPU tests covered now with a mocked implementation of the STOP operation (other emulators seem to mock this one, too, so it should be okay). I'm now looking for a way to integrate the test ROMs into the build pipeline which is pretty tricky since it has to run on a host computer. But then we can make sure the tests won't break anymore.

It would be great if you could create another PR for the PPU upgrades. Would love the pull them in!

As for the audio wiring. I used the circuit I described here: https://blazer82.github.io/FT81x_Arduino_Driver/#audio-setup-section

Only difference is: you take each of the audio pins (currently 7, 6, 4) and feed them through a 1k resistor into the rest of the circuit. Everything else stays the same, just 3 resistors at the beginning instead of 1. Hope this helps.

And yes, Teensy (and for that matter Arduino boards) are not meant for proper debugging 😅

**Comment by blazer82** [comment]
@granthaack 
```
// I fried my teensy so I need to remap the buttons
```

won't pull this one in, though 😂 what happened?

Maybe it's easier if I'm just going to copy your code this time.

**Comment by grantbreaksthings** [comment]
I put 5v to the pins. :cry: I have another one on the way. I removed that change, it should be pretty easy to merge in.

I'll give that audio circuit a shot when I get home from work today.

**Comment by blazer82** [comment]
Oh nice, I recently fried one of my display boards with 9V on the audio line 🤷‍♂️

I already pulled your code in and also attributed you as a contributor 😊

**Comment by blazer82** [comment]
@granthaack I mentioned your contribution in [my latest blog post](https://medium.com/@raphaelstaebler/behold-the-breadboard-gameboy-d966249a482e?sk=f7f10d0bb55876e5ce65c477edf57006). Let me know if you want to have anything changed about the mention!

**Comment by grantbreaksthings** [comment]
Thanks for the credit! I'd like to keep contributing to this if that's ok with you, I figured I'd pick a Blargg test and get it to pass this weekend. If not, I can fork this and work on my own.

I've had a long week of programming at work, and the best way to wind down after a week of programming is more programming :laughing:

**Comment by blazer82** [comment]
>I've had a long week of programming at work, and the best way to wind down after a week of programming is more programming

Haha, I know exactly what you mean 😄

>I'd like to keep contributing to this if that's ok with you

Sure, happy to have you! I even considered giving you more access to the repository. Let me know if keeping your fork up to date is too much of a hassle for you.

>I figured I'd pick a Blargg test and get it to pass 

Awesome, because I'm currently working on getting these tests into the build pipeline 😊

**Pull request merged**